File: problems/905-Sort-Array-By-Parity/solution.java
class Solution {
    private int findNext(int[] nums, int idx, boolean flag) {
        int result = -1;

        if (flag) {
            for (int i = idx + 1; i < nums.length; i++) {
                if (nums[i] % 2 == 0) {
                    result = i;
                    break;
                }
            }
        } else {
            for (int i = idx + 1; i < nums.length; i++) {
                if (nums[i] % 2 != 0) {
                    result = i;
                    break;
                }
            }
        }

        return result;
    }

    public int[] sortArrayByParity(int[] nums) {
        int evenIdx = findNext(nums, -1, true), oddIdx = findNext(nums, -1, false);

        if (evenIdx == -1 || oddIdx == -1) {
            return nums;
        }

        while (evenIdx != -1) {
            if (evenIdx > oddIdx) {
                nums[evenIdx] ^= nums[oddIdx];
                nums[oddIdx] ^= nums[evenIdx];
                nums[evenIdx] ^= nums[oddIdx];

                evenIdx = findNext(nums, evenIdx, true);
                oddIdx = findNext(nums, oddIdx, false);
            } else {
                evenIdx = findNext(nums, oddIdx, true);
            }
        }

        return nums;
    }
}

File: problems/905-Sort-Array-By-Parity/solution.go
func sortArrayByParity(nums []int) []int {
    for {
        oddIdx := 0
        for oddIdx < len(nums) && nums[oddIdx]%2 == 0 {
            oddIdx++
        }

        evenIdx := oddIdx
        for evenIdx < len(nums) && nums[evenIdx]%2 != 0 {
            evenIdx++
        }

        if oddIdx == evenIdx || evenIdx == len(nums) {
            break
        } else {
            nums[oddIdx] ^= nums[evenIdx]
            nums[evenIdx] ^= nums[oddIdx]
            nums[oddIdx] ^= nums[evenIdx]
        }
    }

    return nums
}

